### PR TITLE
fix(replay): keep timing and capacity on the selected database version

### PR DESCRIPTION
**Pull request by jasonqinzhou**
## Why and what changed

An omitted backend version can select an obsolete hard-coded database version during KV capacity estimation, rejecting predictions before replay starts. Direct replay also applied deployment-level version pins after Python capacity preflight: requesting `next` could estimate capacity using `current`, then execute timing using `next`.

Resolve omitted capacity versions from the matching database's maintained `current` slot, preserving the existing latest-version behavior for legacy custom roots. Propagate authored deployment, role, and timing pins before capacity preflight; honor both supported role version aliases. Apply the same maintained-version and effective-root lookup at the analytical EPD execution boundary. Explicit pins and conflict validation remain in effect. Null version aliases behave as omitted, and slot aliases resolve against the effective role-specific systems root.

## Review map

- Risk level: medium; version selection and replay preflight ordering.
- Start with `capacity.py` for database resolution and role aliases, then `_materialize_engine_role` in `runner.py` for pin propagation and `config/epd.py` for analytical EPD normalization. Regressions are in `tests/test_aic.py`, `tests/test_runner.py`, and `tests/test_epd_cli.py`.
- Public or serialized contract changed: no signature, schema, or output-format change. Omitted versions now follow the maintained database default, and authored pins apply consistently to capacity and timing.
- Compatibility or rollback concern: custom roots reach both resolution and estimation; explicit unsupported old pins still fail. Reverting restores the obsolete-default and late-pin bugs.
- Native guard: `resolve_role_timing` calls `AicTimingModel::build` before `materialize_aic_capacity`. Construction copies the canonical estimator's resolved version into `AicTimingConfig`, so normal native replay reaches capacity with the timing provenance already resolved. Real nested replay tests cover this path.

## Evidence

Validated head: `af45f23cd71718b94e04501afc1450107dc50e75`; base: `c8ba25e8ff9d7c17dbb3d980e4c2311071e0c2bd`. Squash-merged into main as `7836e7b0c25899a3b4dd7cbe5a96a2972374de17` on 2026-09-19 at 03:58:32 UTC. The final diff remains six Python source/test files.

- Final-head local tests: **275 passed** with the native extension rebuilt from the integrated Rust tree (macOS/arm64, Python 3.12):
  ```sh
  PYTHONPATH=python/aisimulate/src PYTEST_DISABLE_PLUGIN_AUTOLOAD=1 python -m pytest -c pytest.ini -p pytest_asyncio.plugin tests/test_aic.py tests/test_runner.py tests/test_epd_cli.py -q
  ```
- Integration coverage: **594 passed** at the preceding integrated head `371cfc6c`, before the small selected-root correction. This includes host-offload, CLI config, Sweeper estimator/request controls, and unified traffic runtime coverage. The changed capacity/runner/EPD paths were then rerun on the final head as the 275 checks above.
- New root regression: a canonical timing configuration with `[empty_custom_root, "default"]` selected the usable bundled root, but capacity reverted to the first authored root and failed. Current main's explicit selected-root path returned 27,205 blocks for Qwen/Qwen3-32B-FP8/H200/TP2. The fix preserves canonical timing provenance. Three real replay cases failed before correction and pass afterward for omitted, `next`, and literal pins; actual Python/native database versions match and caller input remains unchanged.
- Ruff lint/format and `git diff --check` pass on the final head.
- Direct replay regression command: `python -m pytest tests/test_runner.py -k direct_replay_capacity_uses_the_resolved_timing_version -q`. Tests execute a real request for flat, canonical, and nested inputs with omitted, `next`, and literal non-current versions, recording actual loaded database versions in Python and native phases and checking input immutability.
- Fast CI: [PR run 35419266971](https://github.com/ai-dynamo/aisimulate/actions/runs/35419266971) passed on the current head; codeowners and DCO also pass.
- Trusted merge-gate CI: [Fast CI 35419492282](https://github.com/ai-dynamo/aisimulate/actions/runs/35419492282) and [Full CI 35419492506](https://github.com/ai-dynamo/aisimulate/actions/runs/35419492506) both passed after current-head review and exact-SHA admission. Full CI completed **45 successful jobs** with five scope-declared N/A components (Rust, Rust public API, Rust feature modes, Cargo Deny and Collector Data) plus the expected PR release-staging skip. The selector explicitly marked each skipped validation component false, and the required aggregate passed. The trusted `pull-request/291` copy was verified equal to the current head. Prior-head CI is historical evidence only.
- CodeRabbit reviewed commit: `af45f23cd71718b94e04501afc1450107dc50e75`; [explicit review](https://github.com/ai-dynamo/aisimulate/pull/291#issuecomment-5739086507) found no actionable findings.
- Codex reviewed commit: `af45f23cd71718b94e04501afc1450107dc50e75`; complete final diff, main integration, native construction guard, version/root propagation, EPD normalization, historical outside-diff findings and tests reviewed; no remaining actionable finding.
- Negative or boundary cases: explicit aliases/literals, missing databases, fixed capacity, custom-root forwarding, duplicate/conflicting versions, and caller input immutability.
- Expected-value derivation: implicit selection is compared against the pre-existing explicit-current path, which bypasses the changed default resolver. Capacities equal that reference, and previously successful paths retain their capacities. This verifies selection, not the accuracy of the memory model itself.
- Before/after direct replay: on `63f3ce0`, requesting `next` in flat replay used `0.24.0` for Python capacity and `0.27.1` for native timing/capacity. Canonical replay instead rejected the prematurely resolved current version. After correction, both phases load `0.27.1`; omitted versions load `0.24.0`. The six-case initial regression reported 2 failures before correction and all corresponding cases pass afterward.
- Original reproduction coverage: Qwen/Qwen3-32B-FP8/H200/TP2 fixed-timing prediction previously rejects `0.19.0`; the fixed CLI completes four requests.

## Modeling or data provenance

No formulas, tables, or goldens changed. Counts below are modeled per-rank KV blocks from the bundled database, not measured performance. Configuration: Qwen/Qwen3-32B-FP8, H200, TP=2, context length 4096. Block sizes are vLLM 64, SGLang 1, and TensorRT-LLM 32 tokens/block, so raw block counts are not cross-backend performance comparisons.

The nine comparisons below were measured at `fad69895` against base `8e763d4e` with the current native extension; the subsequent null/custom-root fixes have separate regression coverage above. Independently authored slot metadata determines the expected version. Each implicit post-fix result matches explicit-current results on both revisions; all capacities are positive and previously successful results are unchanged.

<details>
<summary>Machine-readable before/after capacity results</summary>

```json
[
  {"backend": "vllm", "timing": "default", "version": "0.24.0", "before": 27205, "after": 27205, "explicit_current": 27205},
  {"backend": "vllm", "timing": "fixed", "version": "0.24.0", "before": "version-slot rejection", "after": 27205, "explicit_current": 27205},
  {"backend": "vllm", "timing": "polynomial", "version": "0.24.0", "before": "version-slot rejection", "after": 27205, "explicit_current": 27205},
  {"backend": "sglang", "timing": "default", "version": "0.5.14", "before": 1705736, "after": 1705736, "explicit_current": 1705736},
  {"backend": "sglang", "timing": "fixed", "version": "0.5.14", "before": "version-slot rejection", "after": 1705736, "explicit_current": 1705736},
  {"backend": "sglang", "timing": "polynomial", "version": "0.5.14", "before": "version-slot rejection", "after": 1705736, "explicit_current": 1705736},
  {"backend": "trtllm", "timing": "default", "version": "1.3.0rc20", "before": 55468, "after": 55468, "explicit_current": 55468},
  {"backend": "trtllm", "timing": "fixed", "version": "1.3.0rc20", "before": "version-slot rejection", "after": 55468, "explicit_current": 55468},
  {"backend": "trtllm", "timing": "polynomial", "version": "1.3.0rc20", "before": "version-slot rejection", "after": 55468, "explicit_current": 55468}
]
```

</details>

<details>
<summary>Reproduce the numerical comparison</summary>

Save the script below as `/tmp/capacity-version-evidence.py` and run from the repository root with dependencies and the current native extension installed:

```sh
PYTHONPATH=python/aisimulate/src python /tmp/capacity-version-evidence.py 8e763d4edfd8555240157c2794aa61a27d73b8f8 /tmp/capacity-before.json
PYTHONPATH=python/aisimulate/src python /tmp/capacity-version-evidence.py fad6989501ca89aee19cf6893edb2571087c89c1 /tmp/capacity-after.json
```

Each process loads the selected revision's capacity adapter while retaining the same compiler, native extension, config, and data.

```python
import json
import subprocess
import sys
from pathlib import Path

from aisimulate import capacity

revision = sys.argv[1]
source = subprocess.check_output([
    'git', 'show', f'{revision}:python/aisimulate/src/aisimulate/capacity.py'
], text=True)
exec(compile(source, f'{revision}:capacity.py', 'exec'), capacity.__dict__)

from aisimulate.compiler import prediction_to_replay_spec
from aisimulate.config import CorePredictionConfig
from aisimulate_core.sdk import perf_database

original_get_database = perf_database.get_database
versions = []
def recorded_database(*args, **kwargs):
    database = original_get_database(*args, **kwargs)
    versions.append(database.version)
    return database
perf_database.get_database = recorded_database

rows = []
for backend in ('vllm', 'sglang', 'trtllm'):
    current = perf_database.get_version_slots('h200_sxm', backend)['current']
    for timing in ('default', 'fixed', 'polynomial'):
        timing_config = {'type': timing}
        if timing == 'fixed':
            timing_config.update(prefill_ms=1, decode_ms=1)
        engine = {
            'model': 'Qwen/Qwen3-32B-FP8', 'hardware': 'h200_sxm',
            'backend': backend, 'context_length': 4096,
            'workers': {'aggregated': {'parallelism': {'tensor': 2}, 'timing': timing_config}},
        }
        row = {'backend': backend, 'timing': timing, 'current_slot': current}
        for label, options in [('implicit', {}), ('explicit_current', {'backend_version': current})]:
            versions.clear()
            try:
                config = CorePredictionConfig.model_validate({'engine': dict(engine, **options)})
                deployment = prediction_to_replay_spec(config).backend_deployment
                args = capacity.materialize_aic_num_gpu_blocks(deployment.agg_engine_args)
                row[label] = {'blocks': args['num_gpu_blocks'], 'versions': sorted(set(versions))}
            except Exception as exc:
                row[label] = {'error': str(exc)}
        rows.append(row)
Path(sys.argv[2]).write_text(json.dumps({'revision': revision, 'rows': rows}, indent=2) + '\n')
```

</details>

## Tracking

- Closes: none.
- Related report: [Slack reproduction](https://nvidia.slack.com/archives/C096945HE8M/p1789760962432459).
- [Version-consistency concern](https://github.com/ai-dynamo/aisimulate/pull/291#discussion_r4050828118): correction and actual Python/native version regression coverage verified on the current head; the addressed thread is resolved. The separate changes-requested review was dismissed by the maintainer before this closeout. GitHub recorded APPROVED; the retained human approval was applicable, all four review threads are resolved, and strict Fast CI, Full CI, codeowners and current-main requirements passed before protected squash auto-merge completed. CodeRabbit edge cases and requested boundary coverage remain included.

## Merge verification

Fetched main and verified `7836e7b0c25899a3b4dd7cbe5a96a2972374de17` is on it. Its parent is the validated base `c8ba25e8ff9d7c17dbb3d980e4c2311071e0c2bd`; its tree exactly equals the validated head's `798777dbd6e1f0085e4f61e0f0ecec4d7e3665bf`. The head-to-merge diff is empty.

### CODE REVIEW

**Review by dreamtalen — APPROVED** [review]

**Comment by jasonqinzhou** [comment]
Adapting this approved fix to `main` as requested. The canonical default-timing path on main already resolves the current database version, but fixed/polynomial timing with default KV capacity still reaches the stale Python fallback versions. I reproduced that remaining case with Qwen/Qwen3-32B-FP8 on H200: compilation fails on vLLM `0.19.0` before execution.

The main-branch adaptation will replace those obsolete fallback literals with the maintained `current` alias and carry regression coverage for both the canonical timing path and custom timing with database-backed capacity. Explicit version pins and custom systems roots stay intact. I will integrate current main without rewriting the PR branch history and verify that the final diff contains only the version-default fix and its tests.

**Review by simone-chen — APPROVED** [review]
Superseded by my [Request changes review](https://github.com/ai-dynamo/aisimulate/pull/291#pullrequestreview-5252770989). The version-consistency concern is blocking and must be addressed and the review thread resolved before merge.

This earlier review covered `9cd81ab5`. The PR has since changed to `63f3ce0` and now targets main; the latest review acknowledges the capacity fix and identifies the remaining concerns. My original approval and permission to defer no longer represent my review decision.

File: python/aisimulate/src/aisimulate/capacity.py [hunk 1]
```diff
@@ -234,6 +234,15 @@ def estimate_num_gpu_blocks(
         estimate_num_gpu_blocks as aic_estimate_num_gpu_blocks,
     )
 
+    if backend_version is None:
+        from aisimulate_core.sdk.perf_database import get_latest_database_version
+
+        # Use the maintained current slot (or the latest version in a custom
+        # legacy root), matching the database used for this capacity estimate.
+        backend_version = get_latest_database_version(system, backend_name, systems_paths=systems_path)
```
Comment (simone-chen) [inline python/aisimulate/src/aisimulate/capacity.py:273]: [medium] Resolve the remaining version-consistency concern before merge

This shared capacity lookup addresses part of the earlier concern. However, at this head, `AicTimingConfig::resolved_backend_version` in `crates/core/src/python.rs` still supplies hard-coded defaults such as vLLM `0.19.0`, and `_materialize_engine_role` still materializes capacity before applying `deployment_backend_version` to timing. Resolving an omitted capacity version here alone does not establish that capacity and timing use the same version, especially when a direct replay caller supplies a non-current deployment-level pin without a matching per-role pin.

Please reconcile these paths and add regression coverage for direct replay with an omitted version and a non-current explicit deployment version, checking the versions actually consumed by both capacity and timing. If an apparent path is unreachable or intentionally unsupported, please document the guard and provide the corresponding test evidence. Address this concern and resolve this thread before merging; any remaining deferral needs explicit reviewer agreement. The prior reproductions were on 9cd81ab5; these observations on 63f3ce0 are source-inspected, not a new runtime test result.

**Review by simone-chen — DISMISSED** [review]
This supersedes my earlier approval and its non-blocking wording. The version-consistency concern must be addressed and the review thread resolved before merge. The new head fixes the Python capacity default and updates the stale assertion; the thread below focuses on the remaining cross-entry-point consistency questions. My earlier runtime reproductions were on 9cd81ab5, not this updated head.

File: python/aisimulate/src/aisimulate/capacity.py (hunk 1 above)
Comment (jasonqinzhou) [reply python/aisimulate/src/aisimulate/capacity.py:273]: Confirmed on 63f3ce0: direct flat replay with deployment `backend_version="next"` and no per-role pin estimates Python capacity using `0.24.0`, then native capacity using `0.27.1`. I am fixing propagation before the Python capacity preflight and adding real replay regressions for omitted and non-current deployment versions across flat, canonical, and nested inputs.

The native fallback observation is guarded on current main: `resolve_role_timing` calls `AicTimingModel::build` before `materialize_aic_capacity`, and `build` copies the canonical estimator's resolved backend version into `AicTimingConfig`. The new regressions record versions received by the real Python and native memory estimators and execute a request, covering that ordering without relying on the older release reproduction. I will leave this thread open for your re-review after the correction is pushed.

Comment (jasonqinzhou) [reply python/aisimulate/src/aisimulate/capacity.py:273]: Fixed in ef8feeda, now on PR head fad6989501ca89aee19cf6893edb2571087c89c1 after integrating current main. Authored deployment/role/timing versions are propagated before Python capacity preflight; the capacity adapter also honors the supported `backend_version` role alias. Existing conflict checks remain, and caller dictionaries are copied before adding a timing pin.

The nine real replay cases pass: flat, canonical, and nested inputs, each with omitted version, `next`, or the literal non-current slot version. They execute a request and record actual loaded database versions in the Python and native phases. Both phases use `0.24.0` when omitted and `0.27.1` when pinned. Nested replay covers the native guard: canonical timing construction resolves its provenance before native capacity runs, so the stale fallback is not reached on this main path. Before the correction, the flat non-current case consumed current for Python capacity and next for native capacity, and the canonical case rejected its own prematurely resolved current version.

`tests/test_aic.py` also checks both role version aliases. The focused new checks report 11 passed. Broader current-main tests and fresh CI are running. Please re-review this revision and resolve the concern if satisfied; I have left your thread and review decision intact.

File: python/aisimulate/src/aisimulate/capacity.py [hunk 2]
```diff
@@ -175,7 +175,7 @@ def finish_lowering(value):
         gpu_memory_utilization=lowered.get("gpu_memory_utilization"),
         mem_fraction_static=lowered.get("mem_fraction_static"),
         free_gpu_memory_fraction=lowered.get("free_gpu_memory_fraction"),
-        backend_version=lowered.get("aic_backend_version"),
+        backend_version=lowered.get("aic_backend_version", lowered.get("backend_version")),
```
[the remark this answers is not in the document]
Comment (jasonqinzhou) [reply python/aisimulate/src/aisimulate/capacity.py]: Confirmed. The direct capacity adapter must treat a null legacy alias as omitted so the canonical pin survives. I also reproduced the related replay boundary case: a flat role containing `aic_backend_version: null` is rejected after capacity preflight, and the null key prevents a deployment pin from being propagated. I am normalizing null version aliases as omitted before replay preflight and extending the real omitted/next/literal-version tests to cover that serialized form.

[the remark this answers is not in the document]
Comment (jasonqinzhou) [reply python/aisimulate/src/aisimulate/capacity.py]: Fixed in d2bcd85e75180123d6676fcda7b65dbf475f6d85. Null `aic_backend_version` now falls back to the concrete canonical alias, and flat replay removes null version aliases before propagating a deployment pin. The outside-diff custom-root finding is also fixed: alias resolution falls back to the role's effective `systems_path` before timing overrides are merged.

Before correction, the added null-alias and custom-root regressions both failed; the custom-root case rejected bundled `0.24.0` against custom `current=0.25.0`. Seven targeted checks now pass, including a real request loading `0.25.0` consistently from the custom root and direct replay with omitted, next, and literal pins plus a serialized null field. The broader suite passed 437 cases before adding the three null-replay cases; the full 440-case run is now in progress on this head. The PR description now lists all four changed files and the current review/test state.

File: python/aisimulate/src/aisimulate/capacity.py (hunk 1 above)
Comment (jasonqinzhou) [reply python/aisimulate/src/aisimulate/capacity.py:273]: Final exact-head evidence for `d2bcd85e75180123d6676fcda7b65dbf475f6d85`:

- 440 focused regression tests passed with the native extension rebuilt from current main and this head.
- Direct replay exercises flat, serialized-null, canonical, and nested inputs with omitted, `next`, and literal non-current versions. Python and native database loads agree on `0.24.0` when omitted and `0.27.1` when pinned.
- The custom-root replay uses a deliberately different `current` mapping (`0.25.0` versus bundled `0.24.0`) and loads only `0.25.0` across capacity and timing.
- Null `aic_backend_version` is treated as omitted without mutating the caller's input, so it no longer hides a deployment-level pin.
- Fast CI, DCO, and CODEOWNERS pass on this exact head. The final fresh four-file review found no remaining actionable issue.

Your changes-requested decision and thread remain intact. Please re-review this exact head and resolve the thread if the version-consistency concern is satisfied; Full CI will run after current-head CodeRabbit completes.

**Comment by jasonqinzhou** [comment]
CodeRabbit's current-head review identified one remaining custom-root path. I validated the finding in the source: legacy external AIC timing configurations without canonical estimator controls do not copy `timing_model.config.systems_paths` into the flat role before Python capacity preflight. Capacity can therefore consult the default root while native timing uses the configured root.

I am adding a real replay regression where timing declares both `systems_paths=[custom]` and `systems_path=default`, with different `current` mappings. The correction will preserve native precedence (`systems_paths` first), normalize the configured roots, and make Python capacity use the same first effective root before version resolution.

**Comment by jasonqinzhou** [comment]
Self-review found one boundary issue in the pending custom timing-root fix: it currently resolves timing roots before the existing explicit-capacity and non-AIC early returns. I will move that lookup to the actual estimation path so authored `num_gpu_blocks` values keep their current no-preflight behavior, while inferred capacity still uses the same effective root as timing.

**Comment by jasonqinzhou** [comment]
Final self-review found one precedence edge case in the pending timing-root fix: `systems_paths: []` is explicit and native timing still gives it precedence over `systems_path`, resolving through the default roots. The pending capacity code used truthiness, which could fall back to the singular path. I will switch these checks to `is None` / `is not None` so capacity and timing preserve identical precedence for empty and non-empty lists.

**Comment by jasonqinzhou** [comment]
Follow-up on the empty-list self-review note: the regression test showed the native constructor treats `systems_paths: []` as empty and falls back to `systems_path`. Making the empty list authoritative caused alias resolution to use bundled `0.24.0` while native timing opened the custom root whose current slot is `0.25.0`. I am retaining truthiness-based fallback so the Python preflight matches the observed native contract; the non-empty `systems_paths` precedence case remains covered.

File: python/aisimulate/src/aisimulate/capacity.py (hunk 1 above)
Comment (jasonqinzhou) [reply python/aisimulate/src/aisimulate/capacity.py:273]: Final update on `854519df`: the direct-replay version-consistency fix remains intact, and the latest commit also makes timing-only custom roots consistent during Python capacity preflight. The exact final tree passes 441 focused tests, including real flat/canonical/nested replay, omitted/alias/literal versions, distinct bundled/custom slot mappings, caller immutability, and explicit-capacity early return. Re-review requested; I have left this thread for you to resolve.

**Comment by jasonqinzhou** [comment]
Fast CI on `854519df` found one formatting-only failure in `tests/test_runner.py`: the repository Ruff configuration formats the new parameterized test signature on one line, while the local precheck had used Ruff without the repository config. The source and tests were otherwise green. I reproduced the exact CI command and will apply only that formatter change before rerunning exact-head checks.

**Comment by jasonqinzhou** [comment]
@coderabbitai review\n\nPlease review final head `c6a21d08e83881694244a30d4f7114aabed48a09`. The latest change fixes the verified timing-only custom-root mismatch and adds a real replay regression; the follow-up commit is formatter-only. Exact-head Fast CI and 441 focused tests pass.

**Comment by jasonqinzhou** [comment]
Verified the CodeRabbit cross-layer warning on final head: an EPD prediction with omitted `engine.backend_version` reaches `config/epd.py::_language_execution`, which substitutes SGLang `0.5.10` and fails because the maintained slots are `current=0.5.14` and `next=0.5.17`. I will replace that stale fallback with the same maintained/current database lookup used by capacity, forward the effective timing roots, and add an omitted-version EPD regression. Explicit versions remain resolved unchanged.

**Comment by jasonqinzhou** [comment]
The expanded suite found a boundary regression in the pending EPD fix: standalone omitted versions should resolve to current, but `validate_epd_prediction_mapping` must still reject a callback that deletes an already selected candidate version. Normalizing both sides to current made that deletion invisible. I will add an explicit dropped-selected-version guard while continuing to accept equivalent explicit aliases such as `current`.

File: python/aisimulate/src/aisimulate/capacity.py (hunk 1 above)
Comment (jasonqinzhou) [reply python/aisimulate/src/aisimulate/capacity.py:273]: Final-head update: `6749d4d0` includes the direct-replay version-consistency fix, the timing-only custom-root correction, the EPD omitted-version correction, and current `main` (`bf6dcad9`). The exact merged head passes 444 focused tests and repository Ruff checks. Exact-head Fast CI and CodeRabbit are being refreshed; I have left this thread for you to resolve.

**Comment by jasonqinzhou** [comment]
@coderabbitai review\n\nPlease review final up-to-date head `6749d4d0bc8a7aa6055bfa793eaa6b0565e6db0e`. It fixes the verified EPD cross-layer warning, preserves callback version-deletion validation, and merges current `main`. The exact head passes 444 focused tests and repository Ruff checks.

File: tests/test_epd_cli.py [hunk 3]
```diff
@@ -138,6 +138,56 @@ def test_native_cli_epd_recommend_yaml_predict(tmp_path, capsys, mode, relative_
...
+        "timing_model": {
+            "type": "external",
+            "provider": "aic",
+            "config": {
+                "model": raw["engine"]["model"],
+                "system": raw["engine"]["hardware"],
+                "backend": raw["engine"]["backend"],
+                "tp": 1,
+                "attention_dp": 1,
+                "systems_paths": [str(tmp_path)],
```
[the remark this answers is not in the document]
Comment (jasonqinzhou) [reply tests/test_epd_cli.py]: Confirmed. The implementation handles both branches, but the current tests cover only plural `systems_paths` and successful lookup. I will parameterize the custom-root regression across plural and singular root spellings and add an omitted-version test that forces `get_latest_database_version` to return `None` and asserts the clear `ValueError`.

[the remark this answers is not in the document]
Comment (jasonqinzhou) [reply tests/test_epd_cli.py]: Fixed in `2607508550d6159b766393794877445748141684`. The custom-root regression now runs for both `systems_paths` and singular `systems_path`; a separate regression forces latest-version lookup to return `None` and asserts the `no perf database` error. Five focused EPD cases and the full 446-test suite pass, with exact repository Ruff checks green.

**Comment by jasonqinzhou** [comment]
@coderabbitai review\n\nPlease review final test-only head `2607508550d6159b766393794877445748141684`. It adds the singular-root and missing-database coverage requested in your unresolved thread; five focused EPD tests and 446 focused-suite tests pass.

File: python/aisimulate/src/aisimulate/capacity.py (hunk 1 above)
Comment (jasonqinzhou) [reply python/aisimulate/src/aisimulate/capacity.py:273]: Final exact-head update for re-review: the fix is preserved at `b4075b105863af79ef2980addd060b1f782101f1` after merging current `main` (`2d9e024f`). Omitted versions resolve the maintained `current` slot, explicit pins are preserved, and capacity/timing now consume the same resolved version across direct replay and analytical EPD paths, including custom roots. The current merged-base suite passed 453 tests with the rebuilt native extension; repository Ruff passed; standalone Fast CI passed: https://github.com/ai-dynamo/aisimulate/actions/runs/35410855335; Full CI passed all 55 jobs: https://github.com/ai-dynamo/aisimulate/actions/runs/35410916697; CodeRabbit is green. The single base-merge conflict retained both the version/root lookup and main’s new estimator-policy normalization. I am leaving this thread for reviewer resolution.

Comment (jasonqinzhou) [reply python/aisimulate/src/aisimulate/capacity.py:273]: Guarded final-head update: `main` advanced to `7cfa4f20`, so I disabled the prior auto-merge, merged that base, resolved the two overlaps by preserving both this PR’s version/root behavior and the new engine/request controls, rebuilt the native extension, and requalified exact head `ad520d99ea4e0364b5ba16f005a98d06b0f38552`. The merged-base suite passed 543 tests; repository Ruff passed; standalone Fast CI passed: https://github.com/ai-dynamo/aisimulate/actions/runs/35414018418; Full CI passed all 55 jobs: https://github.com/ai-dynamo/aisimulate/actions/runs/35414069408; CodeRabbit, DCO, and codeowners are green. The PR diff remains the same six intended files. SHA-guarded squash auto-merge is armed for this head; this review/thread remains the sole protected gate and is left for reviewer resolution.

File: python/aisimulate/src/aisimulate/capacity.py [hunk 4]
```diff
@@ -161,6 +165,16 @@ def finish_lowering(value):
...
 
+    capacity_systems_path = lowered.get("systems_path")
+    if timing_system_roots:
+        from .sweeper.forward_pass_estimator import resolve_systems_paths
+
+        resolved_roots = resolve_systems_paths(timing_system_roots)
+        if resolved_roots:
+            # Native timing gives systems_paths precedence over systems_path;
+            # capacity preflight must query that same first effective root.
+            capacity_systems_path = resolved_roots[0]
```
Comment (jasonqinzhou) [inline python/aisimulate/src/aisimulate/capacity.py:176]: [P2] Preserve the root selected by canonical timing construction. With a canonical op_level timing config whose systems_paths is [an existing empty custom directory, "default"], best_available correctly selects the packaged root and lowering records it in systems_path. This block then overwrites that resolved root with the first authored root. Reproduced with Qwen/Qwen3-32B-FP8, H200 and TP2: timing selects the packaged database, but this PR raises an unsupported-model/backend/GPU capacity error; current main returns 27205 blocks. Use the resolved timing provenance when it exists instead of replacing it with the authored search order, and add a real multi-root capacity/replay regression.

**Comment by jasonqinzhou** [comment]
@coderabbitai review

Please review current head af45f23c after current-main integration and the canonical selected-root correction. The new real replay cases reproduce the prior failure with multiple roots for omitted, next, and literal version pins and now pass.

File: python/aisimulate/src/aisimulate/capacity.py (hunk 4 above)
Comment (jasonqinzhou) [reply python/aisimulate/src/aisimulate/capacity.py:176]: Fixed in af45f23c. Canonical timing provenance now retains its selected database root during capacity preflight. Three real replay regressions with an empty first root and a usable later root failed before correction and pass afterward; they cover omitted, next, and literal pins, actual Python/native database loads, and caller immutability. Ruff lint/format and diff checks also pass. The broader suite is still running.

**Review by jasonqinzhou — COMMENTED** [review]
Fresh Codex review of af45f23cd71718b94e04501afc1450107dc50e75 against main c8ba25e8ff9d7c17dbb3d980e4c2311071e0c2bd: no remaining actionable finding.

Reviewed the complete six-file feature diff, capacity/timing consumers, deployment/role/version conflict checks, canonical selected-root provenance, analytical EPD normalization, null aliases, explicit capacity behavior, input immutability, and all historical inline/outside-diff findings. Native replay constructs canonical timing and records its resolved version/root before native capacity estimation; the earlier hard-coded fallback cannot override that resolved provenance on this path.

The newly reproduced P2 root regression is fixed in this head and its thread is resolved. Three new real replay cases failed before the fix and passed afterward. The final affected capacity/runner/EPD suites passed 275 tests with the rebuilt native extension. The preceding integration head passed 594 expanded checks before the small selected-root correction; those broader results are integration evidence, not a claim that all 594 were rerun after the fix. Current-head Ruff lint/format and diff checks passed.

CodeRabbit explicitly reviewed this head with no actionable findings: https://github.com/ai-dynamo/aisimulate/pull/291#issuecomment-5739086507 . PR Fast CI 35419266971 passed; trusted merge-gate Full CI will now be refreshed. Human approval and branch protections remain separate merge gates.

File: python/aisimulate/src/aisimulate/capacity.py (hunk 1 above)
Comment (jasonqinzhou) [reply python/aisimulate/src/aisimulate/capacity.py:273]: Verified the version-consistency correction on current head af45f23cd71718b94e04501afc1450107dc50e75 after integrating main c8ba25e8. Authored deployment pins reach capacity before timing, and actual Python/native database loads agree for flat, null-flat, canonical, nested, and canonical multi-root replay with omitted, next, and literal pins. Canonical timing construction resolves provenance before native capacity. The additional selected-root issue found during closeout is now fixed as well.

The final capacity/runner/EPD suites passed 275 tests with the rebuilt native extension, including all 15 direct replay version cases and custom-root cases. CodeRabbit's explicit current-head review found no actionable findings. The original requested correction and tests are present, so I am resolving this addressed conversation using GitHub's author resolution permission. This resolves the technical conversation; it does not submit or impersonate reviewer approval.

**Comment by jasonqinzhou** [comment]
Verified squash merge into main: `7836e7b0c25899a3b4dd7cbe5a96a2972374de17` (2026-09-19 03:58:32 UTC).

- Validated head: `af45f23cd71718b94e04501afc1450107dc50e75`, including the canonical selected-root correction and three before/after replay regressions.
- Final-head local evidence: 275 capacity/runner/EPD tests passed with the rebuilt native extension; lint/format and diff checks passed. The 594 expanded checks are correctly recorded as preceding-integration-head evidence.
- Exact-head [trusted Fast CI](https://github.com/ai-dynamo/aisimulate/actions/runs/35419492282) and [trusted Full CI](https://github.com/ai-dynamo/aisimulate/actions/runs/35419492506) passed. Full CI: 45 successful jobs; five validation components explicitly marked N/A by the scope selector; PR release staging appropriately skipped. Required PR Fast CI and codeowners also passed.
- Applicable human approval is recorded, all four conversations are resolved, and both current-head technical reviews found no actionable issue.
- Fetched main and verified merge ancestry. The squash commit has parent `c8ba25e8ff9d7c17dbb3d980e4c2311071e0c2bd` and tree `798777dbd6e1f0085e4f61e0f0ecec4d7e3665bf`, exactly matching the tested head; the head-to-merge diff is empty.

PR description and validation notes now reflect the completed merge.
